dbms/src/Formats/ProtobufReader.h

```diff
@@ -97,10 +97,19 @@ class ProtobufReader : private boost::noncopyable
         bool readUInt(UInt64 & value);
         template<typename T> bool readFixed(T & value);
         bool readStringInto(PaddedPODArray<UInt8> & str);
-        bool ALWAYS_INLINE maybeCanReadValue() const { return field_end != REACHED_END; }
+
+        bool ALWAYS_INLINE maybeCanReadValue() const
+        {
+            if (field_end == REACHED_END)
+                return false;
+            if (cursor < root_message_end)
+                return true;
+
+            throwUnknownFormat();
```

Comment:
@bopohaa Can you provide an example of the case when the `ProtobufReader` didn't throw an exception while it was reading invalid data?

Reply:
> 
> 
> @bopohaa I am not sure if this code is useful because I think there are enough checks for the end of the message in the functions `SimpleReader::read*` and we don't need one more check. Have you met a case when the `ProtobufReader` didn't throw an exception while it was reading invalid data?

SimpleReader :: read * reads the entire stream without splitting it into separate messages. Therefore, it can read for quite a long time if the messages arrive with sufficient intensity.

Reply:
Now I understand your point. I agree it's better to detect errors as soon as possible. I've slightly changed your solution because it's faster to only compare `cursor` with `field_end` here (https://github.com/yandex/ClickHouse/pull/6132)